### PR TITLE
Add multiprocess support to snabb-softwire-v2

### PR DESCRIPTION
**Pull request by tsyesika**
This PR addresses the work in #820.

## Schema

It adds a new list element to the configuration allowing the configuration of multiple instances of the lwAFTR using the single configuration file. It also adds the configuration ability of specifying RSS (receive side scaling) queues when supported in the lwAFTR. The configuration changes allow for specifying specific properties in instance specific `external-interface` and `internal-interface` blocks as well as specifying the PCI device. It also allows for configuration of global properties across instances.

It also finally includes the state counters per instance as well as globally. This will allow one to get the counters across all instances (as aggregate) and instance specific to aid debugging and monitoring of the lwAFTR instances.

The new list block is as follows:
```yang
list instance {
  key "device";

  leaf device { type string; description "PCI device"; }

  list queue {
    key "id";

    leaf id { type uint8; }

    container external-interface {
    	leaf ip { type inet:ipv4-address;  mandatory true; }
    	leaf device { type string; }
    	leaf mac { type yang:mac-address; mandatory true; }

    	uses vlan-tagging;

    	container next-hop {
    		choice address {
    			mandatory true;
    			case ip {
    				leaf ip {  type inet:ipv4-address; }
    				leaf resolved-mac {  config false; type yang:mac-address; }
    			}
    			case mac {
    				leaf mac {
    					type yang:mac-address;
    				}
    			}
    		}
    	}
    }

    container internal-interface {
    	leaf ip { type inet:ipv6-address; mandatory true; }
    	leaf mac { type yang:mac-address; mandatory true; }

    	uses vlan-tagging;

    	container next-hop {
    		choice address {
    			mandatory true;
    			case ip {
    				leaf ip { type inet:ipv6-address; }
    				leaf resolved-mac { config false; type yang:mac-address; }
    			}
    			case mac {
    				leaf mac { type yang:mac-address; }
    			}
    		}
    	}
    }
  }

  uses state-counters;
}
```
_The schema above has had the descriptions removed or reduced for ease of reading_

## Migration

The migration requires additional information which is not in older versions of the config (i.e. the PCI device). This is fixed with the introduction of the `-o` or `--option` flag where one specifies an X-Path value to describe the new value. For example

```
$ snabb lwaftr migrate-configuration -f 3.3.0 -o "/softwire-config/instance[device=00:02.0]" my-config.conf
```

The migration converts it so that the new config has a single item in the instance list and a single RSS queue which holds the instance specific values. Additional instances can then be added manually. There is no option to take multiple configurations and convert them to a single configuration with multiple instances.

## Lwaftr changes

The Lwaftr in this PR still lacks multi-process support. This PR is only to add support for the configuration which will allow multi-process support later. The way the lwAFTR deals with this at current is it will error if provided with a configuration which has multiple instances.

The PR tries to keep changes to `apps.lwaftr` with the idea that very little for the specific instances has actually changed and handling multiple instances will be the job of the leader or the setup code. The setup code produces multiple versions of the configuration each with a single instance and provides them to the LwAFTR app.

### Future thinking

When multiprocess support is added, it is expected that the setup code produces multiple single instance configs for each instance specified.

E.g. For a config with three instances. The setup code would produce three configurations each with just one entry in `/softwire-config/instance` and pass that to `apps.lwaftr.LwAftr` which will then internally extract the interfaces and merge them with the global interface producing a config which resembles the config prior to this change.

## Testing

Testing caused a problem in that the tests must be able to be run on multiple servers or on different PCI devices. This previously was not a problem as it was passed in via flags however now it's in the configuration it caused a problem.

The solution I came up with was to migrate the test data to have a device of "test". This is then on converted JIT by the tests to be a configuration with the values that have been specified in `SNABB_PCI0` and `SNABB_PCI1`. If those have not been specified it does not migrate them as it does not use the PCI addresses.

To do this I added a user facing migration which is additional to the `3.3.0` migration discussed above. It is called `pci-device` and it migrates the PCI devices specified in a configuration. The migration takes in three parameters, these are in provided using `-o` or `--option`:

- `from[device=PCI DEVICE]` this is whatever exists in `/softwire-config/instance/device` 
- `internal[device=PCI DEVICE]` this is the new value for `/softwire-config/instance/device`
- `external[device=PCI DEVICE]` this is the new value for `/softwire-config/instance/queue/external-interface/device` if it is unspecified then the option is left empty.

This allows for both the tests to build JIT configurations.

### CODE REVIEW

File: src/lib/yang/snabb-softwire-v2.yang [hunk 1]
```diff
@@ -26,13 +21,117 @@ module snabb-softwire-v2 {
...
+        These configuration options will only affect the specific lwaftr
+        with the given name specified in the name leaf. The other options
+        not present in this list are shared amongst all instances.";
+
+      key "name";
+
+      leaf name {
+        type string;
+        description
+         "Name of lwAFTR instance. This must be unique amongst the Snabb
```
Comment (wingo) [inline src/lib/yang/snabb-softwire-v2.yang]: Here I think we want one name for the whole tree of lwaftr processes, because it will be a single endpoint from a "snabb config" perspective.

File: src/lib/yang/snabb-softwire-v2.yang [hunk 2]
```diff
@@ -26,13 +21,117 @@ module snabb-softwire-v2 {
...
+          mandatory true;
+          description
+           "L3 Address of the internet-facing network interface.  Used
+            when generating error messages and responding to ICMP echo
+            requests.";
+        }
+
+        leaf mac {
+          type yang:mac-address;
+          mandatory true;
```
Comment (wingo) [inline src/lib/yang/snabb-softwire-v2.yang]: Ideally we will be able to pull this from the device itself in the future, FWIW.  The older 82599 driver didn't allow this but the newer one does, I think.

**Review by wingo — COMMENTED** [review]
Looking good.  I think instead of identifying the instances by name though I would identify them by PCI address and queue number.  For now queue number should just be zero I think, until we get RSS support.  WDYT?

File: src/lib/yang/snabb-softwire-v2.yang [hunk 3]
```diff
@@ -242,6 +532,131 @@ module snabb-softwire-v2 {
...
+       "Provides configuration for specific instances of the lwAFTR.
+        These configuration options will only affect the specific lwaftr
+        with the given name specified in the name leaf. The other options
+        not present in this list are shared amongst all instances.";
+
+      key "device";
+
+      leaf device {
+        type string;
+        description "PCI device name.";
```
Comment (wingo) [inline src/lib/yang/snabb-softwire-v2.yang]: Probably needs more description here; i.e. that if no "external-interface" device is specified, that this device is for on-a-stick operation; and that if "external-interface" is specified that this interface will be for internal v6 traffic only; right?

It's a bit of a shame to have to document this in such detail; would be nice if we could figure out a way to make the configuration a bit more self-explanatory.  Maybe we should key on some abstract instance identifier instead and then just have all relevant instance-specific configuration be inside the thing :/

File: src/lib/yang/snabb-softwire-v2.yang [hunk 4]
```diff
@@ -242,6 +532,131 @@ module snabb-softwire-v2 {
...
+                }
+              }
+              case mac {
+                leaf mac {
+                type yang:mac-address;
+                description
+                "Statically configured MAC address of the next hop for the
+                  internet-facing NIC.";
+              }
+              }
```
Comment (wingo) [inline src/lib/yang/snabb-softwire-v2.yang]: I think the body of the `leaf mac` needs indenting here.

File: src/lib/yang/snabb-softwire-v2.yang [hunk 5]
```diff
@@ -242,6 +532,131 @@ module snabb-softwire-v2 {
...
+              mandatory true;
+              case ip {
+                leaf ip {
+                  type inet:ipv4-address;
+                  description
+                  "IPv4 address of the next hop for the internet-facing NIC.
+                    The lwAFTR will resolve this to a MAC address using ARP.";
+                }
+                leaf mac {
+                  config false;
```
Comment (wingo) [inline src/lib/yang/snabb-softwire-v2.yang]: so this would only be returned by get-state?  sgtm.

File: src/lib/yang/snabb-softwire-v2.yang [hunk 6]
```diff
@@ -242,6 +532,131 @@ module snabb-softwire-v2 {
...
+                  description "Resolved next-hop mac address found by ARP."
+                  type yang:mac-address;
+                }
+              }
+              case mac {
+                leaf mac {
+                type yang:mac-address;
```
Comment (wingo) [inline src/lib/yang/snabb-softwire-v2.yang]: indent body

File: src/lib/yang/snabb-softwire-v2.yang [hunk 7]
```diff
@@ -254,9 +669,15 @@ module snabb-softwire-v2 {
          "Maximum packet size to send on the IPv4 interface.";
       }
 
+      leaf mru {
+        type uint16;
+        default 1460;
+        description
+         "Maximm packet size to recieve on the IPv4 interface.";
```
Comment (wingo) [inline src/lib/yang/snabb-softwire-v2.yang]: typo; sgtm though.  does the data plane do mru handling yet?

File: src/lib/yang/snabb-softwire-v2.yang [hunk 8]
```diff
@@ -274,6 +695,13 @@ module snabb-softwire-v2 {
          "Maximum packet size to sent on the IPv6 interface.";
       }
 
+      leaf mru {
+        type uint16;
+        default 1460;
+        description
+         "Maximm packet size to recieve on the IPv6 interface.";
```
Comment (wingo) [inline src/lib/yang/snabb-softwire-v2.yang]: typo also

File: src/lib/yang/snabb-softwire-v2.yang [hunk 9]
```diff
@@ -392,374 +820,25 @@ module snabb-softwire-v2 {
           }
         }
       }
+
+      container binding-table-version {
```
Comment (wingo) [inline src/lib/yang/snabb-softwire-v2.yang]: This will end up being a bit verbose: `/snabb-softwire/binding-table/binding-table-version/binding-table-version`.  What about removing `binding-table-`?  Also is the time zone defined for the date/time?

I worry a bit about feature creep here; I wonder if we can punt on some of these.  Dunno.

**Review by wingo — COMMENTED** [review]
Thanks for this branch!  Some notes.

The obvious one is that although good it's a bit large and hard to review.  That might be unavoidable, but I think it means that we have to put extra effort into making it reviewable.  Firstly if the PR isn't a WIP, can you remove the WIP tag? Secondly I think it makes sense to update the PR summary with a general idea of what's changed, not only for this review but also as a hint towards later readers.  In particular I think it should be possible to collect a list of user-visible changes from the summaries of merged PRs.  Of course updating CHANGELOG.md in the PR is also nice but not always needed.

Speaking of which, this PR has user-visible changes, so we need to update documentation as well.

Given that we have the ability to migrate a configuration and re-setting the PCI addresses, does that mean that we could actually keep the old --v4 / --v6 arguments?  If passed, they could cause migration of the config file given on the command line.  Alternately, should we include some arbitrary XPath-based replacement thing?  E.g. "use `foo.conf`, but set `/softwire-config/instance` to `{ device 03:00.0; queue { id 1; }}`".

What is the NUMA story?  Before, NUMA PCI device affinity would be checked before running.  Is that still the case?

What's missing for next steps?

File: src/apps/lwaftr/lwaftr.lua [hunk 10]
```diff
@@ -238,12 +238,29 @@ local function init_transmit_icmpv4_reply (rate_limiting)
...
+function select_instance(conf, device)
+   -- Merges t1 into t2 by mutating t1 with t2's values. This is quicker than
+   -- producing a new table with the combination of t1 & t2 like I'd have liked.
+   local function table_merge(t1, t2)
+      for k,v in pairs(t2) do t1[k] = v end
+   end
+
+   instance = assert(conf.softwire_config.instance[device])
+
+   table_merge(conf.softwire_config.external_interface,
```
Comment (wingo) [inline src/apps/lwaftr/lwaftr.lua]: I am not sure that we should be mutating the incoming configuration, dunno.  We should copy it if we need to mutate it IMO.

File: src/program/lwaftr/run_nohw/run_nohw.lua [hunk 11]
```diff
@@ -62,9 +62,12 @@ function run(parameters)
    local conf_file, b4_if, inet_if, opts = parse_args(parameters)
    local conf = require('apps.lwaftr.conf').load_lwaftr_config(conf_file)
    local c = config.new()
+   local lwaftr_class =  LwAftr
+   lwaftr_class.config_arg = "conf"
```
Comment (wingo) [inline src/program/lwaftr/run_nohw/run_nohw.lua]: I think we should avoid mutating the LwAftr class.

File: src/program/lwaftr/run_nohw/run_nohw.lua [hunk 12]
```diff
@@ -62,9 +62,12 @@ function run(parameters)
...
+   local device = next(assert(conf.softwire_config.instance))
 
    -- AFTR
-   config.app(c, "aftr", LwAftr, conf)
+   config.app(c, "aftr", lwaftr_class, {conf=conf, device=device})
```
Comment (wingo) [inline src/program/lwaftr/run_nohw/run_nohw.lua]: Hummmmmmmmmmmm.  This seems like a perilous change to make.  Right now the configuration of the lwaftr app is described by the snabb-softwire-v2 schema, and that's how it gets shipped over the wire.  Making it ad-hoc is going in the wrong direction, I think....

Comment (wingo) [reply src/program/lwaftr/run_nohw/run_nohw.lua]: What if instead the lwaftr app just gets a normal snabb-softwire-v2 config, but with only one device?  I.e. computing the lwaftr app's config from the whole machine's config will produce a config with no "instance" section, and the "internal-interface" / "external-interface" sections have been specialized for the device at hand.  WDYT?

Comment (tsyesika) [reply src/program/lwaftr/run_nohw/run_nohw.lua]: Yeah.

I've changed it so that's how it works now.

File: src/program/lwaftr/run_nohw/run_nohw.lua (hunk 11 above)
Comment (tsyesika) [reply src/program/lwaftr/run_nohw/run_nohw.lua]: Fixed.

**Comment by tsyesika** [comment]
I've updated the documentation and all the tests pass. This still removes the  `--v4` and `--v6` parameters expecting the user to manually migrate the configuration. We could include them however there are a number of reasons why I might consider not doing so:

1. When we do come to add true multiprocess support it'll be ambiguous what to do if those options are specified and the configuration has multiple instances configured (do we migrate one of them to the new addresses provided, are these addresses suppose to specify a single, already configured, instance which should run, etc.)
2. This is not an issue if we're just generating a temporary file with the v4 and v6 addresses from the command line specified but I'd be very much against modifying the actual file in-case something were to go wrong mid migration, we currently have no tooling to verify the migration occurred successfully. (aside from checking status codes or the likes).

I also noticed that our release numbers and migration configuration versions are now widely out of sync. Prior to this PR the v2 migrations in master thus far assume a version of `3.2.0` and this one with more configuration backwards incompatibilities builds on that with `3.3.0` however our latest release to date I believe is `3.1.8`. I think this should be solved, it seems there are two main ways to do this:

1. Change the version numbers in the migration to match with the actual version numbers.
2. Next release bump the version number to 3.3.0 to match the migrations.

I have no real preference which however I was bumping the 3.X.0 number as configuration changes require action to be able to use the lwaftr and so are somewhat major releases.

File: src/apps/lwaftr/lwutil.lua [hunk 13]
```diff
@@ -6,6 +6,8 @@ local S = require("syscall")
 local bit = require("bit")
 local ffi = require("ffi")
 local lib = require("core.lib")
+local corelib = require("core.lib")
```
Comment (wingo) [inline src/apps/lwaftr/lwutil.lua]: Duplicate import

File: src/apps/lwaftr/lwutil.lua [hunk 14]
```diff
@@ -18,6 +20,19 @@ local ehs = constants.ethernet_header_size
...
 local ntohs = lib.ntohs
 
+-- Produces configuration for each instance.
+-- Provided a multi-process configuration it will iterate over each instance
+-- and produce a configuration for each device with a single instance in. This
+-- is then able to be provided to the lwaftr app.
+function produce_instance_configs(conf)
+   local ret = {}
+   for device, queues in pairs(conf.softwire_config.instance) do
+      ret[device] = corelib.deepcopy(conf)
```
Comment (wingo) [inline src/apps/lwaftr/lwutil.lua]: Change to just use `lib`

File: src/lib/yang/snabb-softwire-v2.yang (hunk 8 above)
Comment (wingo) [reply src/lib/yang/snabb-softwire-v2.yang]: Still a typo here

File: src/program/lwaftr/doc/benchmarking.md [hunk 15]
```diff
@@ -10,12 +10,20 @@ example assumes NICs `02:00.0` and `02:00.1` and wired to NICs `02:00.0` and
...
+    src/program/lwaftr/tests/data/icmp_on_fail.conf >> /tmp/icmp_on_fail.conf
+````
+
 In one server, start the lwAFTR:
 
 ```
 $ sudo numactl -m 0 taskset -c 1 ./src/snabb lwaftr run -v \
-    --conf program/lwaftr/tests/data/icmp_on_fail.conf \
-    --v4 0000:02:00.0 --v6 0000:02:00.1
+    --conf /tmp/icmp_on_fail.conf
```
Comment (wingo) [inline src/program/lwaftr/doc/benchmarking.md]: About this.  You mention that you're hesitant to keep supporting --v4 and --v6 given that we're looking to support multiple processes.  However given that we run the same softwire configuration all the time over different interfaces, and we have code that can migrate files in memory, it would certainly make sense to keep supporting --v4 / --v6 in single-interface mode; and in multiple-interface mode, you could specify `--v4=03:00.0,03:00.1` for example and that can make sense.  What do you think here?

Comment (tsyesika) [reply src/program/lwaftr/doc/benchmarking.md]: Sounds reasonable. I'll add this in to work like you've described.

File: src/apps/lwaftr/lwaftr.lua [hunk 16]
```diff
@@ -8,6 +8,7 @@ local lwdebug = require("apps.lwaftr.lwdebug")
 local lwheader = require("apps.lwaftr.lwheader")
 local lwutil = require("apps.lwaftr.lwutil")
 
+local cltable = require("lib.cltable")
```
Comment (wingo) [inline src/apps/lwaftr/lwaftr.lua]: Stale import, perhaps?

File: src/apps/lwaftr/lwutil.lua [hunk 17]
```diff
@@ -7,6 +7,7 @@ local bit = require("bit")
 local ffi = require("ffi")
 local lib = require("core.lib")
 
+
```
Comment (wingo) [inline src/apps/lwaftr/lwutil.lua]: Spurious extra line here

File: src/apps/lwaftr/lwutil.lua [hunk 18]
```diff
@@ -18,6 +19,19 @@ local ehs = constants.ethernet_header_size
 local o_ipv4_flags = constants.o_ipv4_flags
 local ntohs = lib.ntohs
 
+-- Produces configuration for each instance.
```
Comment (wingo) [inline src/apps/lwaftr/lwutil.lua:21]: Can we put this in `program.lwaftr.setup` instead?  This lwutils file is a little weird and I'd like to avoid adding to it.

File: src/program/lwaftr/doc/filters-performance.md [hunk 19]
```diff
@@ -48,7 +48,7 @@ $ cat ~/bin/run-lwaftr
 BASEDIR=/home/kbarone/snabbswitch/src/
 CONF="`realpath $1`"
 cd ${BASEDIR} && sudo numactl -m 0 taskset -c 1 \
- ./snabb lwaftr run --conf ${CONF} --v4-pci 0000:02:00.0 --v6-pci 0000:02:00.1
+ ./snabb lwaftr run --conf ${CONF}
```
Comment (wingo) [inline src/program/lwaftr/doc/filters-performance.md]: Probably need to revert this change

File: src/program/lwaftr/doc/running.md [hunk 20]
```diff
@@ -69,9 +81,7 @@ just one NIC, because traffic is not symmetric. To do this, use --on-a-stick,
 and specify only one PCI address:
 
 ```
-$ sudo ./snabb lwaftr run \
-    --conf program/lwaftr/tests/data/icmp_on_fail.conf \
-    --on-a-stick 0000:02:00.1
+$ sudo ./snabb lwaftr run --conf /tmp/icmp_on_fail.conf \
```
Comment (wingo) [inline src/program/lwaftr/doc/running.md]: Consider revert of this change

File: src/program/lwaftr/run/run.lua [hunk 21]
```diff
@@ -95,11 +105,11 @@ function parse_args(args)
    function handlers.reconfigurable() opts.reconfigurable = true end
    function handlers.h() show_usage(0) end
    lib.dogetopt(args, handlers, "b:c:vD:yhir:n:",
-      { conf = "c", v4 = 1, v6 = 1, ["v4-pci"] = 1, ["v6-pci"] = 1,
-        verbose = "v", duration = "D", help = "h", virtio = "i", cpu = 1,
-        ["ring-buffer-size"] = "r", ["real-time"] = 0, ["bench-file"] = "b",
-        ["ingress-drop-monitor"] = 1, ["on-a-stick"] = 1, mirror = 1,
-        hydra = "y", reconfigurable = 0, name="n" })
+   { conf = "c", v4 = 1, v6 = 1, ["v4-pci"] = 1, ["v6-pci"] = 1,
```
Comment (wingo) [inline src/program/lwaftr/run/run.lua]: I'm really sorry to be picky but could we possibly undo this indentation change?  Otherwise it looks like line 108 is a new statement, to me at least.

File: src/program/lwaftr/run_nohw/run_nohw.lua (hunk 11 above)
Comment (wingo) [reply src/program/lwaftr/run_nohw/run_nohw.lua]: Can you take another look here?  It looks like we're still mutating the class.

File: src/program/lwaftr/run_nohw/run_nohw.lua (hunk 12 above)
Comment (wingo) [reply src/program/lwaftr/run_nohw/run_nohw.lua]: Looks like you might have updated another front-end (`run` perhaps) but forgot this one.  is that possible?

File: src/program/lwaftr/setup.lua [hunk 22]
```diff
@@ -18,23 +19,55 @@ local ipv6_apps  = require("apps.lwaftr.ipv6_apps")
...
 end
 
-function lwaftr_app(c, conf)
+-- Checks the existance and NUMA affinity of PCI devices
+-- NB: "nil" can be passed in and will be siliently ignored.
+local function validate_pci_devices(devices)
+   for _, address in pairs(devices) do
+      assert(lwutil.nic_exists(address),
+             ("Could not locate PCI device '%s'"):format(address))
+      numa.check_affinity_for_pci_addresses(address)
```
Comment (wingo) [inline src/program/lwaftr/setup.lua]: I think this `check_affinity_for_pci_addresses` takes an array of addresses; can you verify?

File: src/program/lwaftr/setup.lua [hunk 23]
```diff
@@ -180,41 +219,46 @@ local function link_sink(c, v4_out, v6_out)
...
 
 function load_on_a_stick(c, conf, args)
-   lwaftr_app(c, conf)
-   local v4_nic_name, v6_nic_name, v4v6, pciaddr, mirror = args.v4_nic_name,
-      args.v6_nic_name, args.v4v6, args.pciaddr, args.mirror
-   local external_interface = conf.softwire_config.external_interface
-   local internal_interface = conf.softwire_config.internal_interface
+   local inst_configs = lwutil.produce_instance_configs(conf)
+   local device, lwaftr_config = next(inst_configs)
+   local queue = lwaftr_config.softwire_config.instance.test.queue.values[1]
```
Comment (wingo) [inline src/program/lwaftr/setup.lua]: Is "test" the right thing here?

File: src/program/lwaftr/tests/data/icmp_endaddr.conf [hunk 24]
```diff
@@ -42,27 +42,37 @@ softwire-config {
...
     mtu 2000;
-    next-hop {
-      mac 68:68:68:68:68:68;
-    }
     reassembly {
       max-fragments-per-packet 40;
     }
   }
+  instance {
+    device test;
```
Comment (wingo) [inline src/program/lwaftr/tests/data/icmp_endaddr.conf:51]: In a followup I think we should change this.  I think I pushed you into making what in hindsight looks like a wrong decision, to make the PCI device be the "key" to identify an instance and then have queues inside that.  Having to specify a PCI address when there is simply no sensible value as in here is a bit wonky.  Additionally code that does e.g.

```lua
+   local inst_configs = lwutil.produce_instance_configs(conf)
+   local device, lwaftr_config = next(inst_configs)
+   local queue = lwaftr_config.softwire_config.instance[device].queue.values[1]
```

I find this really weird to use the noun "queue" to reflect what is really a lwAFTR instance.  I know how it comes about but I think that reflects that we're organizing things wrong; perhaps we should have instead

```yang
list instance {
   key name;
   leaf name { type string; }
   leaf device { type string; }
   leaf queue { type uint8; default 0; }
   container external-interface { ... }
   ...
}
```

This allows us to refer to instances simply by name.  If we need a PCI address or queue number, that's up to `setup.lua` to assert.  Otherwise it's possible to run without these, as in `bench` etc...  Dunno.  Definitely followup material though.  Already having done what you have to this point is good, if we decide to migrate again (wdyt about that?) we can do that in a followup.

File: src/program/lwaftr/tests/subcommands/monitor_test.py [hunk 25]
```diff
@@ -23,7 +24,7 @@ class TestMonitor(BaseTestCase):
         str(SNABB_CMD), 'lwaftr', 'run',
         '--name', DAEMON_PROC_NAME,
         '--bench-file', '/dev/null',
-        '--conf', str(DATA_DIR / 'icmp_on_fail.conf'),
+        '--conf', CONFIG_PATH,
```
Comment (wingo) [inline src/program/lwaftr/tests/subcommands/monitor_test.py]: Do we still need to make these changes if the `snabb lwaftr run` is handling setting the PCI devices on incoming configurations?

**Review by wingo — COMMENTED** [review]
Overall looking very good, some comments.  I am done with the full review, let's address nits and get this merged :)

File: src/program/lwaftr/tests/data/icmp_endaddr.conf (hunk 24 above)
Comment (tsyesika) [reply src/program/lwaftr/tests/data/icmp_endaddr.conf:51]: We actually discussed using `name` as the key a while ago and decided against it. I think if I remember correctly we didn't want to do that because the name is not a per instance name but rather per collection of instances. For example you'd want to be able to do:

```
$ snabb config get mylwaftr /
```

This should find the collection of instances with the name `mylwaftr` and displaying all of them, if that was the key in an instance surely it'd only display that single instance. I also don't know if we're wanting specific `internal-interface` and `external-interface` settings per queue (VLAN tag or whatever) but with the schema above that wouldn't be possible. What would be the use of the `queue` list in the above schema?

File: src/program/lwaftr/tests/selftest.sh [hunk 26]
```diff
@@ -13,6 +13,15 @@ else
     TEST_WHAT="*"
 fi
 
+# Create a directory for JIT migrated configs so that each configuration file
```
Comment (wingo) [inline src/program/lwaftr/tests/selftest.sh]: Can we remove this and all the bits about generating files with the specific PCI addresses in them?

**Review by wingo — APPROVED** [review]
lgtm, you are a hero!!!!
